### DISCUSSION TITLE
picows websocket client

### DISCUSSION THREAD

**Discussion by Tapanhaz** [tarasko/picows#1] [Q&A]
I tried implementing picows for connecting to a broker websocket. https://github.com/Tapanhaz/Shoonya-Websocket

**Comment by tarasko** [comment]
Thank you for trying out picows! 

I quickly ran through the code in on_ws_frame and noticed a couple of thing that you maybe would find interesting to change:

* I would advice to add 
```
if frame.msg_type == WSMsgType.PING:
     transport.send_pong(frame.get_payload_as_bytes())
```
to make sure you client always replies to ping messages.

* Also, it is likely that Shoonya never fragments messages, I believe it is very common not to fragment messages for brokers.  It does not have much practical benefits. It only make sense to have this logic 
```
if frame.fin:
    ...
else:
    ...
```
if you have actually observed non fin messages.  Imho, it may actually suffice to just have this instead:
```
def on_ws_frame(self, transport, frame):
    assert frame.fin, "unexpected fragmented websocket message from Shoonya"
    if frame.msg_type == WSMsgType.PING:
         transport.send_pong(frame.get_payload_as_bytes())
    elif frame.msg_type == WSMsgType.CLOSE:
         # It is good to maybe log close reason and close message here and disconnect
         # logger.info("Shoonya disconnected, code=%s, reason=%s, frame.get_close_code(), frame.get_close_message().decode())
         transport.disconnect()
    else:
        assert frame.msg_type == WSMsgType.TEXT, f"Shoonya is expected to send text messages, instead received {frame.msg_type}"
        msg = frame.get_payload_as_utf8_text()
        self.parent.on_data_callback(msg)
```
Notice that I also used get_payload_as_utf8_text instead of get_payload_as_bytes, as it is faster than first getting bytes and then decoding them with decode('utf8')

I wish all the best with your project!

**Reply by Tapanhaz** [reply]
Thank you for the guidance. Yes, Shoonya's messages are not fragmented. I'll update the code in on_ws_frame as suggested. We need to send a custom ping message {"t": "h"} to the server every 3 seconds. Since the server doesn't respond to pings, I believe I can safely eliminate the ping message check in on_ws_frame. I'll test this and update you on the results. Thanks again!

**Reply by Tapanhaz** [reply]
I was wrong. The server sends pong .., modified my code.  Maybe I made some huge mistakes there.  With low number of subscriptions it performs quite well but when token numbers increased, it is lagging. i checked with websocket-client (running in asyncio.to_thread) side by side, put a counter in on_ticks to check  the number of ticks it gets per sec / total tokens subscribed to check average ticks per sec (both with same sets of tokens) : 
![Screenshot from 2024-09-06 11-02-49](https://github.com/user-attachments/assets/aaf34445-1502-4e66-8cbb-2c006a0d716e)
![Screenshot from 2024-09-06 11-01-05](https://github.com/user-attachments/assets/c492d101-5dbc-4abc-961e-bb535f1635a2)
I am trying to debug at my end but not able to get success till now....

**Reply by tarasko** [reply]
Without profiling it is difficult to say where the bottleneck is, have you tried running it with cProfile? 

Regarding running code in asyncio.to_thread, it most likely wouldn't improve things because of GIL. I would expect it to actually slow down everything unless you are doing some heavy slow IO there. 

> Note Due to the [GIL](https://docs.python.org/3/glossary.html#term-GIL), asyncio.to_thread() can typically only be used to make IO-bound functions non-blocking. However, for extension modules that release the GIL or alternative Python implementations that don’t have one, asyncio.to_thread() can also be used for CPU-bound functions.

One thing that I noticed, that may help a little bit in ShoonyaClient
Instead of 
```
msg = json.loads(msg)
```
Create json decoder once and reuse it 
```
# Somewhere
json_decoder = json.JSONDecoder()
...
# In on_data_callback
        try:
            msg = json_decoder.loads(msg)
```
But again without profiling it is not possible to say where the bottleneck is

**Reply by Tapanhaz** [reply]
I profile it with cProfile and scalene and didn't find anything other than your mentioned line . scalene detects ssl_context.verify_mode = ssl.CERT_REQUIRED and another line in on_ws_frame which I forget to comment out . But as the market is closed as of now, I will paste here the live profiling report on monday.  Also I am thinking of replacing json with orjson.

**Reply by Tapanhaz** [reply]
This is the cprofile result ( running with 252 tokens subscribed for 60 seconds): Please remove the .txt extension

[shoonya_ticker.prof.txt](https://github.com/user-attachments/files/16925781/shoonya_ticker.prof.txt)

**Reply by tarasko** [reply]
Hm, I don't see anything suspicious in the profiler result.

Does ticker messages arrive at constant rate or are they only coming when trade tick happens? Could it be that there was simply not much messages from shoonya during those 60 seconds?

I saw that on_ws_frame was called 1582 times during 60 seconds, which suggest 26 messages (ticks) per seconds, which is not 0.09 but it is a very small number anyway.
Just to give some numbers, I'm running a single threaded app where I connect to a few websockets via picows and sometimes it is more than 10000-20000 messages per second.
 
During those 60 seconds was CPU heavily consumed by the client?

Do you use threads/locks in the client? You mentioned earlier asyncio.to_thread, would it be possible to see what do main and __subscribe_callback provided to ShoonyaTicker do?

I'm asking because I saw 
```
call_soon_threadsafe
```
which suggesting that threading is somehow involved.

I'd guess it is either not much messages from the broker or a deadlock somewhere, probably related to multithreading and synchronization primitives

**Comment by Tapanhaz** [comment]
Extremely sorry, I made a huge mistake in subscription :
![Screenshot from 2024-09-10 10-51-01](https://github.com/user-attachments/assets/a5e3684f-faf3-44a8-914d-45dde30187ab)


Sorry again for the trouble I created because of own silly mistake..

**Reply by Tapanhaz** [reply]
This is scalene report now : 
![Screenshot from 2024-09-10 11-24-00](https://github.com/user-attachments/assets/f34f5b8c-b2ce-409a-9d13-83047b4edb6f)

and this is cprofile result : 

[shoonya_ticker_new.prof.txt](https://github.com/user-attachments/files/16940144/shoonya_ticker_new.prof.txt)

**Reply by Tapanhaz** [reply]
with orjson : 
![Screenshot from 2024-09-10 11-43-24](https://github.com/user-attachments/assets/8b9e4901-4d44-4fe5-9f7b-2d58221354bd)

**Reply by tarasko** [reply]
Cheers! A good luck with your library :)

**Comment by Tapanhaz** [comment]
On network disconnection, it behaves differently in windows and linux. In windows (python 3.11 and 3.12), It disconnects immediately on network disconnects but in linux (Ubuntu 24.04 LTS, python 3.12.3 and 3.12.6) it shows a indefinite behavior. So I make a loop checking for pong after each ping with a timeout of 20 secs:

![Screenshot from 2024-10-06 10-22-31](https://github.com/user-attachments/assets/4394c4d1-aa42-4bb1-bf8e-2127f6a67a65)

If not received, It will disconnects itself and the retry mechanism will try to reconnect it with a gap of 1 sec: 


![Screenshot from 2024-10-06 10-56-43](https://github.com/user-attachments/assets/842261a2-7105-4e21-a4d7-713488765ada)

But transport.disconnect() takes a significant amount of time in between:


![Screenshot from 2024-10-06 10-17-07](https://github.com/user-attachments/assets/7536c47c-46cd-49c3-a759-04d03ccb0320)

**Reply by tarasko** [reply]
May I ask you to try to call 
```
self.transport.underlying_transport.abort()
```
instead of 
```
self.transport.disconnect()
```
And check if it will resolve the issue?

The problem with `disconnect` is that it tries to flush all buffered messages first and potentially wait for confirmation (TCP_ACK) that they were received. It could be that the linux kernel is waiting for TCP_ACK for the ping messages that you sent earlier.

If `abort` helps I will add it to `WSTransport`

**Reply by tarasko** [reply]
`picows` currently doesn't have autoping feature, but I'm planning to add it this week.

**Reply by Tapanhaz** [reply]
Thanks it is working...

![Screenshot from 2024-10-06 20-22-38](https://github.com/user-attachments/assets/25e064a1-9138-4578-9e32-3a9353af1c26)

I really loved current low level minimalistic interface, User gets more fine grained control over each thing.

**Reply by tarasko** [reply]
Awesome! Thanks for the testing!

**Reply by tarasko** [reply]
Hi @Tapanhaz ,
I recently added an automatic ping-pong mechanism for detecting stale connections. This might help to simplify your code a bit.

https://picows.readthedocs.io/en/stable/guides.html#auto-ping-pong

**Reply by Tapanhaz** [reply]
Thanks for the update. That sounds great !! This will definitely help simplify things on my end.

**Reply by Tapanhaz** [reply]
I recently tried implementing  auto ping in  another repo : https://github.com/Tapanhaz/Kite-Ticker . This is my first hands on cython too..
